python/rikai/spark/sql/codegen/mlflow_registry.py

```diff
@@ -0,0 +1,275 @@
+            mlflow.set_tracking_uri(self.tracking_uri)
+            self._artifact = getattr(mlflow, self.flavor).load_model(self.uri)
+        return self._artifact
+
+    def _run_to_spec_dict(self):
+        """Convert the Run into a ModelSpec
+
+        Returns
+        -------
+        spec: Dict[str, Any]
+        """
+        tags = self._run.data.tags
+        spec = {
+            "version": tags["rikai.spec.version"],
+            "schema": tags["rikai.output.schema"],
+            "model": {"flavor": tags["rikai.model.flavor"]},
+        }
+
+        # transforms
+        transforms = {}
+        if "rikai.transforms.pre" in tags:
+            transforms["pre"] = tags["rikai.transforms.pre"]
+        if "rikai.transforms.post" in tags:
+            transforms["post"] = tags["rikai.transforms.post"]
+        if len(transforms) > 0:
+            spec["transforms"] = transforms
+
+        # options
+        options = dict(self._run.data.params or {})  # for training
+        for key, value in tags.items():
+            key = key.lower().strip()
+            if key.startswith("rikai.option."):
+                sub_len = len("rikai.option.")
+                options[key[sub_len:]] = value
+        if len(options) > 0:
+            spec["options"] = options
+
+        # uri (to the model artifact)
+        if "rikai.model.artifact_path" in tags:
+            spec["model"]["uri"] = os.path.join(
+                "runs:/",
+                self._run.info.run_id,
+                tags["rikai.model.artifact_path"],
+            )
+        return spec
+
+
+def codegen_from_run(
+    spark: SparkSession,
+    run: mlflow.entities.Run,
+    tracking_uri: str,
+    name: Optional[str] = None,
+    options: Optional[Dict[str, str]] = None,
+) -> str:
+    """Generate code from an MLFlow runid
+
+    Parameters
+    ----------
+    spark : SparkSession
+        A live spark session
+    run : mlflow.entities.Run
+        the mlflow run that produced the model we want to register
+    name : str
+        The name of the model in the catalog
+    options : dict
+        Optional parameters passed to the model.
+
+    Returns
+    -------
+    str
+        Spark UDF function name for the generated data.
+    """
+    try:
+        spec = MlflowModelSpec(run, options=options, tracking_uri=tracking_uri)
+    except Exception:
+        to_spec_msg = (
+            "Could not create well-formed ModelSpec from run {}."
+            "Check MLFlowModelSpec._run_to_spec_dict"
+        ).format(run_id)
+        raise SpecError(to_spec_msg) from e
+    udf = udf_from_spec(spec)
+    return register_udf(spark, udf, name)
+
+
+def log_model(
+    model: Any,
+    artifact_path: str,
+    flavor: str,
+    schema: str,
+    pre_processing: Optional[str] = None,
+    post_processing: Optional[str] = None,
+    registered_model_name: Optional[str] = None,
+    **kwargs,
+):
+    """Convenience function to log the model with information needed by rikai
```

Comment:
Is this used in training process

could be more explicitly by offering examples?

Reply:
yes during training. yeah i'll add an example.